File: mycodes/98-ValidateBinarySearchTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        boolean isValid = true;
        isValid &= root.left == null ? true : root.left.val < root.val;
        isValid &= root.right == null ? true : root.right.val > root.val;
        isValid = isValid && helper(root.left, null, root) && helper(root.right, root, null);
        return isValid;
    }

    public boolean helper(TreeNode root, TreeNode leftparent, TreeNode rightparent) {
        boolean isValid = true;
        if (root == null || (root.left == null && root.right == null)) return true;
        isValid &= root.left == null ? true : root.left.val < root.val;
        isValid &= root.right == null ? true : root.right.val > root.val;
        isValid &= root.right == null || rightparent == null ? true : root.right.val < rightparent.val;
        isValid &= root.left == null || leftparent == null ? true : root.left.val > leftparent.val;
        isValid = isValid && helper(root.left, leftparent, root) && helper(root.right, root, rightparent);
        return isValid;
    }
}

class Solution2 {
    long pre = Long.MIN_VALUE;
    public boolean isValidBST(TreeNode root) {
        if (root == null) return true;
        boolean isValid = isValidBST(root.left) & root.val > pre;
        pre = root.val;
        return isValid && isValidBST(root.right);
    }
}

File: mycodes/98-ValidateBinarySearchTree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var res []int

func isValidBST(root *TreeNode) bool {
  res = make([]int, 0)
  traverse(root)
  for i := 1; i < len(res); i++ {
    if res[i] <= res[i-1] {
      return false
    }
  }
  return true
}

func traverse(root *TreeNode) {
  if root == nil {
    return
  }
  traverse(root.Left)
  res = append(res, root.Val)
  traverse(root.Right)
}
